pkg/kf/logs/tailer.go

```diff
@@ -0,0 +1,137 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// Tailer reads the logs for a KF application. It should be created via
+// NewTailer().
+type Tailer interface {
+	// Tail tails the logs from a KF application and writes them to the
+	// writer.
+	Tail(ctx context.Context, appName string, out io.Writer, opts ...TailOption) error
+}
+
+type tailer struct {
+	client corev1.CoreV1Interface
+}
+
+// NewTailer creates a new Tailer.
+func NewTailer(client corev1.CoreV1Interface) Tailer {
+	return &tailer{
+		client: client,
+	}
+}
+
+// Tail tails the logs from a KF application and writes them to the writer.
+func (t *tailer) Tail(ctx context.Context, appName string, out io.Writer, opts ...TailOption) error {
+	cfg := TailOptionDefaults().Extend(opts).toConfig()
+	if appName == "" {
+		return errors.New("appName is empty")
+	}
+
+	if cfg.NumberLines < 0 {
+		return errors.New("number of lines must be greater than or equal to 0")
+	}
+
+	namespace := cfg.Namespace
+	logOpts := v1.PodLogOptions{
+		Container: "user-container",
+		Follow:    cfg.Follow,
+	}
+
+	if cfg.NumberLines != 0 {
+		n := int64(cfg.NumberLines)
+		logOpts.TailLines = &(n)
+	}
+
+	if err := t.watchForPods(ctx, namespace, appName, out, logOpts); err != nil {
+		return fmt.Errorf("failed to watch pods: %s", err)
+	}
+	return nil
+}
+
+func (t *tailer) watchForPods(ctx context.Context, namespace, appName string, out io.Writer, opts v1.PodLogOptions) error {
+	w, err := t.client.Pods(namespace).Watch(metav1.ListOptions{
+		LabelSelector: "serving.knative.dev/service=" + appName,
+	})
+	if err != nil {
+		return err
+	}
+	for e := range w.ResultChan() {
+		if e.Type != watch.Added {
+			continue
+		}
+		go func(e watch.Event) {
+			t.readLogs(ctx, e.Object.(*v1.Pod).Name, out, opts)
+			w.Stop()
+		}(e)
+	}
+	return nil
+}
+
+func (t *tailer) readLogs(ctx context.Context, name string, out io.Writer, opts v1.PodLogOptions) {
+	for ctx.Err() == nil {
+		if err := t.readStream(ctx, name, out, opts); err != nil {
+			log.Printf("[WARN] %s", err)
+		}
+
+		if !opts.Follow {
+			return
+		}
+	}
+}
+
+func (t *tailer) readStream(ctx context.Context, name string, out io.Writer, opts v1.PodLogOptions) error {
+	// XXX: This is not tested at a unit level and instead defers to
+	// integration tests.
+	req := t.client.
+		Pods("default").
+		GetLogs(name, &opts).
+		Context(ctx)
+
+	stream, err := req.Stream()
+	if err != nil {
+		return fmt.Errorf("failed to read stream: %s", err)
+	}
+	defer stream.Close()
+
+	buf := make([]byte, 1024)
+	for ctx.Err() == nil {
+		n, err := stream.Read(buf)
```

Comment:
could this use `io.Copy` instead? they handle a couple of additional edge-cases.

Reply:
Thats a great idea!